src/Microsoft.AspNetCore.Routing.Abstractions/Endpoint.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Diagnostics;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    [DebuggerDisplay("{DisplayName,nq}")]
+    public abstract class Endpoint
+    {
+        protected Endpoint(EndpointMetadataCollection metadata, string displayName)
+        {
+            // Both allowed to be null
+            Metadata = metadata ?? EndpointMetadataCollection.Empty;
+            DisplayName = displayName;
+        }
+
+        public string DisplayName { get; }
+
+        public EndpointMetadataCollection Metadata { get; }
+    }
```

Comment:
This class as well as the feature interface will need to go to the HTTP repo once we're slightly more baked. Note that this just has enough info for something like a middleware to introspect metadata. 

The separation between this and `MatcherEndpoint` is intentional and allows middleware to consume this API without a dependency on too many routing concepts.

Reply:
Yes